Python/ceval.c

```diff
@@ -4518,8 +4519,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             PyTypeObject *cls_type = Py_TYPE(cls);
             assert(cls_type->tp_dictoffset > 0);
             PyObject *dict = *(PyObject **) ((char *)cls + cls_type->tp_dictoffset);
-            // Don't care if no dict -- tp_version_tag should catch anything wrong.
-            DEOPT_IF(dict != NULL && ((PyDictObject *)dict)->ma_keys->dk_version !=
+            // All types should have a dict. dk_version is also an identity check
+            // to ensure `cls` is really a type object. Otherwise, reading
+            // tp_version_tag may segfault.
+            DEOPT_IF(dict == NULL || ((PyDictObject *)dict)->ma_keys->dk_version !=
```

Comment:
There is no need to check the version of the keys, the type version is sufficient.

Reply:
Perhaps I was trying too hard, but I had the impression that _PyType_Check was slow, so I'd used the `dk_version` to validate.

Now that I've taken a look at _PyType_Check's code, I realized I was naive to think that way :). It's just a ``type->tp_flags & Py_TPFLAGS_TYPE_SUBCLASS`` once inlined. Plenty faster than a dict calculation.